File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2021/Y21D07.kt
package org.gristle.pdxpuzzles.advent.y2021

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.iteration.minMax
import org.gristle.pdxpuzzles.utilities.parsing.getIntList
import kotlin.math.abs

class Y21D07(input: String) : Day {

    private val crabs: List<Int> = input.getIntList()

    private val crabRange: IntRange = crabs
        .minMax()
        .let { (min, max) -> min..max }

    private tailrec fun List<Int>.optimalAlignmentCost(range: IntRange = crabRange, fuelCost: (Int) -> Int): Int {

        fun List<Int>.alignmentCost(position: Int, fuelCost: (Int) -> Int) =
            sumOf { fuelCost(abs(it - position)) }

        fun IntRange.midPoint() = (last - first) / 2 + first

        if (range.first == range.last) return alignmentCost(range.first, fuelCost)
        val midPoint = range.midPoint()
        val newRange = listOf(range.first..midPoint, midPoint..range.last)
            .minBy { alignmentCost(it.midPoint(), fuelCost) }

        return optimalAlignmentCost(newRange, fuelCost)
    }

    override fun part1() = crabs.optimalAlignmentCost { it }

    override fun part2() = crabs.optimalAlignmentCost { (1..it).sum() }
}

fun main() = Day.runDay(Y21D07::class)

//    Class creation: 7ms
//    Part 1: 343468 (6ms)
//    Part 2: 96086265 (17ms)
//    Total time: 31ms

File: rust/advent/src/bin/2021/y21d07.rs
use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Int = i64;
type Input = (Vec<Int>, Int, Int);
type Output = Int;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(21, 7).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let crabs: Vec<_> = input.get_numbers().collect();
    let (&min, &max) = crabs.iter().minmax().into_option().unwrap();
    (crabs, min, max)
}

fn optimal_alignment_cost<F>(crabs: &Vec<Int>, mut min: Int, mut max: Int, fuel_cost: F) -> Int 
where F: Fn(Int) -> Int,
{
    while min != max {
        let mp = mid_point(min, max);
        let min_cost = alignment_cost(crabs, mid_point(min, mp), &fuel_cost);
        let max_cost = alignment_cost(crabs, mid_point(mp, max), &fuel_cost);
        if min_cost <= max_cost {
            max = mp;
        } else {
            min = mp;
        }
    }
    alignment_cost(crabs, min, &fuel_cost)
}

fn alignment_cost<F>(crabs: &Vec<Int>, position: Int, fuel_cost: &F) -> Int 
where F: Fn(Int) -> Int,
{
    crabs.iter().map(|&crab| fuel_cost((crab - position).abs())).sum()
}

fn mid_point(min: Int, max: Int) -> Int {
    (max - min) / 2 + min
}

fn part1((crabs, min, max): &Input) -> Output {
    optimal_alignment_cost(crabs, *min, *max, |it| it)
}

fn part2((crabs, min, max): &Input) -> Output {
    optimal_alignment_cost(crabs, *min, *max, |it| {
        (1..=it).sum()
    })
}

#[test]
fn default() {
    let input = get_input(21, 7).unwrap();
    let input = parse_input(&input);
    assert_eq!(343468, part1(&input));
    assert_eq!(96086265, part2(&input));
}

// Input parsed (34μs)
// 1. 343468 (10μs)
// 2. 96086265 (32μs)
// Total: 78μs